File: Java/Solutions/Problem006/XorList.java
package Problem006;

import java.util.HashMap;
import java.util.Map;

/**
 * XOR Linked list containing XOR list nodes
 */
public class XorList<T> {
    private final Map<Integer, XORListNode> _memory = new HashMap<Integer, XORListNode>();
    private int _memoryAddress = 1;
    private boolean _hasNodes;

    public void add(T value) {
        if (!_hasNodes) {
            _memory.put(_memoryAddress, new XORListNode(_memoryAddress, value, 0));
            _hasNodes = true;
        } else {
            var currentNode = _memory.get(1);
            var previousNode = (XORListNode) null;

            while(true) {
                var nextNodeAddress = currentNode.getBoth() ^ (previousNode == null ? 0 : previousNode.getAddress());
                if (nextNodeAddress == 0) break;

                previousNode = currentNode;
                currentNode = _memory.get(nextNodeAddress);
            }

            var nodeToAdd = new XORListNode(_memoryAddress, value, currentNode.getAddress());
            currentNode.setBoth(currentNode.getBoth() ^ nodeToAdd.getAddress());
            _memory.put(_memoryAddress, nodeToAdd);
        }
        _memoryAddress++;
    }

    public T get(int index) {
        if (index >= _memory.size())
            throw new IndexOutOfBoundsException("Index out of bound");
        var currentNode = _memory.get(1);
        var previousNode = (XORListNode) null;

        for (var i = 0; i < index; i++) {
            var nextNodeAddress = currentNode.getBoth() ^ (previousNode == null ? 0 : previousNode.getAddress());

            previousNode = currentNode;
            currentNode = _memory.get(nextNodeAddress);
        }

        return  currentNode.getValue();
    }

    /**
     * Node in a XOR linked list
     */
    public class XORListNode {
        private final int address;
        private final T value;
        private int both;

        public XORListNode(int address, T value, int both) {
            this.address = address;
            this.value = value;
            this.both = both;
        }

        public int getAddress() {
            return address;
        }

        public int getBoth() {
            return both;
        }

        public void setBoth(int both) {
            this.both = both;
        }

        public T getValue() {
            return value;
        }
    }
}

File: C#/Solutions/Problem006/XorList.cs
using System;
using System.Collections.Generic;

namespace DailyCodingProblems.Solutions.Problem006
{
    /// <summary>
    /// XOR linked list
    /// </summary>
    /// <typeparam name="T"></typeparam>
    public class XorList<T>
    {
        private readonly IDictionary<int, XorListNode> _memory = new Dictionary<int, XorListNode>();
        private int _memoryAddress = 1;
        private bool _hasNodes;

        public void Add(T value)
        {
            if (!_hasNodes)
            {
                _memory.Add(_memoryAddress, new XorListNode(_memoryAddress, 0, value));
                _hasNodes = true;
            }
            else
            {
                var currentNode = _memory[1];
                var previousNode = (XorListNode)null;

                while (true)
                {
                    var nextNodeAddress = currentNode.Both ^ (previousNode?.Address ?? 0);
                    if (nextNodeAddress == 0) break;

                    previousNode = currentNode;
                    currentNode = _memory[nextNodeAddress];
                }

                var nodeToAdd = new XorListNode(_memoryAddress, currentNode.Address, value);
                currentNode.Both ^= nodeToAdd.Address;
                _memory.Add(_memoryAddress, nodeToAdd);
            }

            _memoryAddress++;
        }

        public T Get(int index)
        {
            if (index >= _memory.Count) throw new IndexOutOfRangeException("Index out of bounds");
            var currentNode = _memory[1];
            var previousNode = (XorListNode)null;

            for (var i = 0; i < index; i++)
            {
                var nextNodeAddress = currentNode.Both ^ (previousNode?.Address ?? 0);

                previousNode = currentNode;
                currentNode = _memory[nextNodeAddress];
            }

            return currentNode.Value;
        }

        /// <summary>
        /// Nodes in a <see cref="XorList{T}"/>
        /// </summary>
        /// <typeparam name="T"></typeparam>
        public class XorListNode
        {
            public int Address { get; }
            public int Both { get; set; }
            public T Value { get; }

            public XorListNode(int address, int both, T value)
            {
                Address = address;
                Both = both;
                Value = value;
            }
        }
    }
}
